compiler/fir/checkers/checkers.native/src/org/jetbrains/kotlin/fir/analysis/native/checkers/FirNativeObjCRefinementChecker.kt

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright 2010-2022 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.native.checkers
+
+import org.jetbrains.kotlin.descriptors.ClassKind
+import org.jetbrains.kotlin.descriptors.annotations.KotlinTarget
+import org.jetbrains.kotlin.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.FirSession
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.checkers.declaration.FirBasicDeclarationChecker
+import org.jetbrains.kotlin.fir.analysis.checkers.getAllowedAnnotationTargets
+import org.jetbrains.kotlin.fir.analysis.checkers.unsubstitutedScope
+import org.jetbrains.kotlin.fir.analysis.diagnostics.native.FirNativeErrors
+import org.jetbrains.kotlin.fir.containingClass
+import org.jetbrains.kotlin.fir.declarations.*
+import org.jetbrains.kotlin.fir.expressions.FirAnnotation
+import org.jetbrains.kotlin.fir.expressions.coneClassLikeType
+import org.jetbrains.kotlin.fir.resolve.toFirRegularClass
+import org.jetbrains.kotlin.fir.resolve.toSymbol
+import org.jetbrains.kotlin.fir.scopes.FirTypeScope
+import org.jetbrains.kotlin.fir.scopes.getDirectOverriddenFunctions
+import org.jetbrains.kotlin.fir.scopes.getDirectOverriddenProperties
+import org.jetbrains.kotlin.fir.symbols.SymbolInternals
+import org.jetbrains.kotlin.fir.symbols.impl.FirCallableSymbol
+import org.jetbrains.kotlin.fir.symbols.impl.FirNamedFunctionSymbol
+import org.jetbrains.kotlin.fir.symbols.impl.FirPropertySymbol
+import org.jetbrains.kotlin.fir.symbols.impl.LookupTagInternals
+import org.jetbrains.kotlin.name.ClassId
+import org.jetbrains.kotlin.name.FqName
+
+@OptIn(SymbolInternals::class, LookupTagInternals::class)
+object FirNativeObjCRefinementChecker : FirBasicDeclarationChecker() {
+
+    private val hidesFromObjCClassId = ClassId.topLevel(FqName("kotlin.native.HidesFromObjC"))
+    private val refinesInSwiftClassId = ClassId.topLevel(FqName("kotlin.native.RefinesInSwift"))
+    private val supportedTargets = arrayOf(KotlinTarget.FUNCTION, KotlinTarget.PROPERTY)
+
+    override fun check(declaration: FirDeclaration, context: CheckerContext, reporter: DiagnosticReporter) {
+        checkAnnotation(declaration, context, reporter)
+        checkDeclarations(declaration, context, reporter)
+        // TODO: checkFakeOverrides(declaration, context, reporter)
```

Comment:
FIR, unlike K1, does not creates fake overrides for any member which has no real override in class, but you still can get all such declarations from scope:
```kotlin
private fun getFakeOverrides(classSymbol: FirClassSymbol<*>, typeScope: FirTypeScope, session: FirSession) {
    val functions = mutableListOf<FirNamedFunctionSymbol>()
    val properties = mutableListOf<FirPropertySymbol>()
    typeScope.processAllFunctions { symbol ->
        if (
            symbol.isSubstitutionOrIntersectionOverride ||
            symbol.dispatchReceiverClassOrNull()?.let { session.symbolProvider.getSymbolByLookupTag(it) } != classSymbol
        ) {
            functions += symbol
        }
    }        
    typeScope.processAllProperties { symbol ->
        if (symbol !is FirPropertySymbol) return@processAllProperties
        if (
            symbol.isSubstitutionOrIntersectionOverride ||
            symbol.dispatchReceiverClassOrNull()?.let { session.symbolProvider.getSymbolByLookupTag(it) } != classSymbol
        ) {
            properties += symbol
        }
    }
}
```

Reply:
Little explanation and examples:
```kotlin
interface A {
    fun foo()
}

interface B {
    fun foo()
}

interface C : A, B {
    // fake-override fun foo()      <---------- intersection override
}

// -------------------------------------------------------------------------------------

interface Base<T> {
    fun foo(): T
}

interface Derived : Base<Int> {
    // fake-override fun foo(): Int      <---------- substitution override
}

// -------------------------------------------------------------------------------------

interface Base {
    fun foo() // (1)
}

interface Derived : Base {
    // no fake-overrides
    // scope of Derived will return symbol (1), same as scope of Base
}
```

Reply:
Awesome. For this checker we just need to check the intersection overrides. All other overrides are already checked by the `FirNativeObjCRefinementChecker`.